source/s3_client.c

```diff
@@ -2054,3 +2054,159 @@ struct aws_byte_cursor aws_s3_meta_request_resume_token_upload_id(
 
     return aws_byte_cursor_from_c_str("");
 }
+
+static uint64_t s_upload_timeout_threshold_ns = 5000000000; /* 5 Secs */
+const size_t g_expect_timeout_offset_ms =
+    700; /* 0.7 Secs. From experienments on c5n.18xlarge machine for 30 GiB upload, it gave us best performance. */
+
+/**
+ * S3 upload part request has a very low rate (around 0.1%) result in a very high time to first byte latency (5
+ * seconds).
+ * This function used to decide what's the good timeout for upload part based on completed upload part requests made
+ * from the client.
+ * This serves as an optimization to set a timeout that will retry on those upload parts response way too slow.
+ *
+ * The current algorithm:
+ * 1. Start without timeout. After 10 requests completed, we know the average of how long the
+ *      request takes. We decide if it's worth to set a timeout or not. (If the average of request takes more than 5
+ *      secs or not)
+ * 2. If it is worth to retry, start a timeout default to 1 sec.
+ * 3. Once a request finishes without timeout, use the average response_to_first_byte_time + g_expect_timeout_offset_ms
+ *      as our expected timeout. (TODO: The real expected timeout should be a P99 of all the requests.)
+ *  3.1 Adjust the current timeout against the expected timeout, via 0.99 * <current timeout> + 0.01 * <expected
+ *      timeout> to get closer to the expected timeout.
+ * 4. Once request finishes with timeout. We check the timeout rate.
+ *  4.1 If timeout rate is larger than 0.1%, we increase the timeout by 100ms (Check the timeout when the
+ *      request was made, if the updated timeout is larger than the expected, skip update).
+ *  4.2 If timeout rate is larger than 1%, we increase the timeout by 1 secs (If needed). And clear the rate
+ *      to get the exact rate with new timeout.
+ *  4.3 Once the timeout is larger than 5 secs, we stop the process.
+ *
+ * Invoked from `s_s3_auto_ranged_put_send_request_finish`.
+ */
+void aws_s3_client_update_upload_part_timeout(
+    struct aws_s3_client *client,
+    struct aws_s3_request *finished_upload_part_request,
+    int finished_error_code) {
+
+    aws_s3_client_lock_synced_data(client);
+    struct aws_s3_upload_part_timeout_stats *stats = &client->synced_data.upload_part_stats;
+    if (stats->stop_timeout) {
+        /* Timeout was disabled */
+        goto unlock;
+    }
+
+    struct aws_s3_request_metrics *metrics = finished_upload_part_request->send_data.metrics;
+    size_t current_timeout_ms = aws_atomic_load_int(&client->upload_timeout_ms);
+    uint64_t current_timeout_ns =
```

Comment:
debatable: immediately convert to ms instead of ns? so that the math and variables are 99% ms, instead of 50/50 ms/ns

Reply:
I am trying to use ns to do the math for those needs to be round up, so that it will be more accurate.

Reply:
then do 99% ns
it's still seems very error-prone, bouncing back and forth between ms and ns constantly

Reply:
I think I am using 99% of ns.

`ms` only used for store the final result and comparing with the the value we stored before.